### PR TITLE
BulkData 2.0.0: Update the OperationDefinitions and CapabilityStatement #3082

### PR DESCRIPTION
**Pull request by prb112**
- Update the OperationDefinitions and CapablityStatement artifacts
- Relax _type cardinality #3081
- Add integration tests

```
    "instantiates": [
        "http://hl7.org/fhir/uv/bulkdata/CapabilityStatement/bulk-data|1.0.1",
    ],
```

should become `2.0.0` when hitting the endpoint. 

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/ExportOperation.java [hunk 1]
```diff
@@ -28,17 +25,16 @@
...
+import com.ibm.fhir.registry.FHIRRegistry;
 import com.ibm.fhir.server.spi.operation.AbstractOperation;
 import com.ibm.fhir.server.spi.operation.FHIROperationContext;
 import com.ibm.fhir.server.spi.operation.FHIRResourceHelpers;
 
 /**
- * <a href="https://hl7.org/Fhir/uv/bulkdata/OperationDefinition-export.html">BulkDataAccess V1.0.0: STU1 -
- * ExportOperation</a> Creates an System Export of FHIR Data to NDJSON format system export operation definition for
- * <code>$export</code>
+ * Creates an System Export of FHIR Data to NDJSON format
```
Comment (tbieste) [inline operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/ExportOperation.java:34]: "Creates a System..."

File: operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/PatientExportOperation.java [hunk 2]
```diff
@@ -1,35 +1,25 @@
...
-import com.ibm.fhir.model.format.Format;
-import com.ibm.fhir.model.parser.FHIRParser;
 import com.ibm.fhir.model.resource.OperationDefinition;
+import com.ibm.fhir.registry.FHIRRegistry;
 
 /**
- * <a href="https://hl7.org/Fhir/uv/bulkdata/OperationDefinition-patient-export.json.html">BulkDataAccess V1.0.0: STU1 -
- * ExportOperation</a> Creates an Patient Export of FHIR Data to NDJSON format patient export operation definition for
- * <code>$export</code>
+ * Creates an Patient Export of FHIR Data to NDJSON
```
Comment (tbieste) [inline operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/PatientExportOperation.java:13]: "Creates a Patient..."

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java [hunk 3]
```diff
@@ -675,6 +675,39 @@ public void testBaseExport() throws Exception {
...
+    public void testBaseExportWithMultipleTypes() throws Exception {
+        if (ON) {
+            Parameters parameters = generateParameters(FORMAT_NDJSON, null, null, null);
+            Entity<Parameters> entity = Entity.entity(parameters, FHIRMediaType.APPLICATION_FHIR_JSON);
+
+            Response response = getWebTarget()
+                                    .path(BASE_VALID_URL)
+                                    .queryParam("_outputFormat", FORMAT_NDJSON)
+                                    .queryParam("_type", "Patient,Coverage")
+                                    .queryParam("_type", "Patient")
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java:688]: seems like you'd want to add a new type here (something other than Patient or Coverage) to prove its doing a union of the values

Comment (prb112) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java:688]: Yes - I can add one of those as well. I wanted to test to make sure we were  not getting 4x results

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java [hunk 4]
```diff
@@ -675,7 +676,7 @@ public void testBaseExport() throws Exception {
         }
     }
 
-    @Test(groups = { TEST_GROUP_NAME }, dependsOnMethods = { "testGroup" })
+    @Test//(groups = { TEST_GROUP_NAME }, dependsOnMethods = { "testGroup" })
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java]: remove the commented part?

File: fhir-server-test/src/test/resources/test.properties [hunk 5]
```diff
@@ -61,4 +61,4 @@ test.bulkdata.import.azure.enabled = false
 test.bulkdata.export.azure.enabled = false
 
 # Commented out - this is added back in during automation.
-# test.bulkdata.path = /liberty-runtime/wlp/usr/servers/fhir-server/output
+#test.bulkdata.path = /liberty-runtime/wlp/usr/servers/fhir-server/output
```
Comment (prb112) [inline fhir-server-test/src/test/resources/test.properties:64]:
```suggestion
# test.bulkdata.path = /liberty-runtime/wlp/usr/servers/fhir-server/output
```

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java (hunk 4 above)
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java]:
```suggestion
    @Test(groups = { TEST_GROUP_NAME }, dependsOnMethods = { "testGroup" })
```

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java [hunk 6]
```diff
@@ -275,12 +275,14 @@ public Response doGet(String path, String mimeType, String provider, String outc
...
         Response response;
+        // We're doing a tight loop here so we can do an immediate DELETE it.
+        ZonedDateTime startTime = ZonedDateTime.now();
         do {
             response = doGet(exportStatusUrl, FHIRMediaType.APPLICATION_FHIR_JSON, "default", "default");
             // 202 accept means the request is still under processing
             // 200 mean export is finished
             assertEquals(Status.Family.familyOf(response.getStatus()), Status.Family.SUCCESSFUL);
-            Thread.sleep(1000);
+            waitForSetPeriodOfTime(startTime, 5, 1000);
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/bulkdata/ExportOperationTest.java:285]: I looked this one up and the second argument is the maximum time to wait in minutes.  So essentially this is saying throw and exception to break us out of this loop if its been five minutes.  Smart.

File: operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/ExportOperation.java [hunk 7]
```diff
@@ -64,10 +57,14 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
...
         // Pick off parameters
+        javax.ws.rs.core.UriInfo uriInfo = (javax.ws.rs.core.UriInfo) operationContext.getProperty(FHIROperationContext.PROPNAME_URI_INFO);
+        MultivaluedMap<String, String> queryParameters = uriInfo.getQueryParameters();
+
         MediaType outputFormat = export.checkAndConvertToMediaType(parameters);
 
         Instant since = export.checkAndExtractSince(parameters);
-        List<String> types = export.checkAndValidateTypes(parameters);
+
+        List<String> types = export.checkAndValidateTypes(parameters, queryParameters);
```
Comment (lmsurpre) [inline operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/ExportOperation.java:67]: I thought that our operations framework automatically converted the query parameters into a Parameters object.
Maybe it just doesn't merge the two if we receive both?

File: operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/util/BulkDataExportUtil.java [hunk 8]
```diff
@@ -199,7 +201,14 @@ public Instant checkAndExtractSince(Parameters parameters) {
...
 
-    public List<String> checkAndValidateTypes(Parameters parameters) throws FHIROperationException {
+    /**
+     * processes both the Parameters object and the query parameters
+     * @param parameters
+     * @param queryParameters
+     * @return
+     * @throws FHIROperationException
+     */
+    public List<String> checkAndValidateTypes(Parameters parameters, MultivaluedMap<String, String> queryParameters) throws FHIROperationException {
```
Comment (lmsurpre) [inline operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/util/BulkDataExportUtil.java:211]: In the case of a GET with queryParameters, I think our operation framework automatically converts that into a Parameters object.
I assume we'd feed both of those into this method?  Assuming thats right, does the redundancy cause any issue?

Comment (prb112) [reply operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/util/BulkDataExportUtil.java:211]: Right, it's redundant for GET, I can add a check - with a post - it does not

Comment (lmsurpre) [reply operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/util/BulkDataExportUtil.java:211]: ok, interesting.  maybe
1. add a test for GET to confirm it behaves how we want
2. i open a new issue to discuss/decide whether the framework should combine query parameters with the Parameters in the body
?

Comment (prb112) [reply operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/util/BulkDataExportUtil.java:211]: Sure. sounds fine. I'll add a test

Comment (prb112) [reply operation/fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/util/BulkDataExportUtil.java:211]: Test is added

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**